game-core/src/main/java/games/strategy/triplea/delegate/power/calculator/AvailableSupports.java

```diff
@@ -0,0 +1,198 @@
+package games.strategy.triplea.delegate.power.calculator;
+
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.attachments.UnitSupportAttachment;
+import games.strategy.triplea.delegate.Matches;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import org.triplea.java.collections.CollectionUtils;
+import org.triplea.java.collections.IntegerMap;
+
+/**
+ * Tracks the available support that a collection of units can give to other units.
+ *
+ * <p>Once a support is used, it will no longer be available for other units to use.
+ */
+@Data
+@Builder(access = AccessLevel.PRIVATE)
+@AllArgsConstructor
+public class AvailableSupports {
+
+  public static final AvailableSupports EMPTY_RESULT =
+      AvailableSupports.builder()
+          .supportRules(new HashMap<>())
+          .supportUnits(new HashMap<>())
+          .build();
+
+  final Map<UnitSupportAttachment.BonusType, List<UnitSupportAttachment>> supportRules;
+  final Map<UnitSupportAttachment, IntegerMap<Unit>> supportUnits;
+
+  /** Sorts 'supportsAvailable' lists based on unit support attachment rules. */
+  static AvailableSupports getSortedSupport(
+      final Collection<Unit> unitsGivingTheSupport,
+      final Set<UnitSupportAttachment> rules,
+      final boolean defence,
+      final boolean allies) {
+    final AvailableSupports supportCalculationResult =
+        getSupport(unitsGivingTheSupport, rules, defence, allies);
+
+    final SupportRuleSort supportRuleSort =
+        SupportRuleSort.builder()
+            .defense(defence)
+            .friendly(allies)
+            .roll(UnitSupportAttachment::getRoll)
+            .strength(UnitSupportAttachment::getStrength)
+            .build();
+    supportCalculationResult
+        .getSupportRules()
+        .forEach((bonusType, unitSupportAttachment) -> unitSupportAttachment.sort(supportRuleSort));
+    return supportCalculationResult;
+  }
+
+  /**
+   * Constructs an AvailableSupportTracker to track the possible support given by the units
+   *
+   * @param defence are the receiving units defending?
+   * @param allies are the receiving units allied to the giving units?
+   */
+  public static AvailableSupports getSupport(
```

Comment:
Codacy found an issue: [The method 'getSupport(Collection, Collection, boolean, boolean)' has an NPath complexity of 1203, current threshold is 200](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=6326866)